### PR TITLE
[Search Registered Models Pagination] order_by in search

### PR DESCRIPTION
**Pull request by ankit-db**
## What changes are proposed in this pull request?

This PR includes an implementation that parses a list of `order_by` clauses that can be passed into `search-registered-models`. 

## How is this patch tested?

Unit testing

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [x] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for
Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [x] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/utils/search_utils.py [hunk 1]
```diff
@@ -317,30 +318,43 @@ def run_matches(run):
...
         is_ascending = True
         if token_value.lower().endswith(" desc"):
             is_ascending = False
             token_value = token_value[0:-len(" desc")]
         elif token_value.lower().endswith(" asc"):
             token_value = token_value[0:-len(" asc")]
+        return token_value, is_ascending
+
+    @classmethod
+    def parse_order_by_runs(cls, order_by):
```
Comment (mparkhe) [inline mlflow/utils/search_utils.py:344]: rename as `parse_order_by_for_search_runs`

File: mlflow/utils/search_utils.py [hunk 2]
```diff
@@ -317,30 +318,43 @@ def run_matches(run):
...
+        token_value, is_ascending = cls._parse_order_by_string(order_by)
         identifier = cls._get_identifier(token_value.strip(), cls.VALID_ORDER_BY_ATTRIBUTE_KEYS)
-        return (identifier["type"], identifier["key"], is_ascending)
+        return identifier["type"], identifier["key"], is_ascending
+
+    @classmethod
+    def parse_order_by_registered_models(cls, order_by):
```
Comment (mparkhe) [inline mlflow/utils/search_utils.py:350]: rename `parse_order_by_for_search_registered_models`
at some point soon (different pull request), we should refactor this as `SearchUtils` base class simpler/shorter method names and derived classes for `SearchRuns`, `SearchModelVersions`, `SearchRegisteredModels`, ... etc.

File: mlflow/utils/search_utils.py [hunk 3]
```diff
@@ -24,6 +24,7 @@ class SearchUtils(object):
         CASE_INSENSITIVE_STRING_COMPARISON_OPERATORS.union({'='})
     VALID_SEARCH_ATTRIBUTE_KEYS = set(RunInfo.get_searchable_attributes())
     VALID_ORDER_BY_ATTRIBUTE_KEYS = set(RunInfo.get_orderable_attributes())
+    VALID_ORDER_BY_KEYS_REGISTERED_MODELS = set(["name", "last_updated_timestamp"])
```
Comment (mparkhe) [inline mlflow/utils/search_utils.py:27]: use existing constants

```
from mlflow.store.model_registry.dbmodels.models import SqlRegisteredModel
VALID_ORDER_BY_KEYS_REGISTERED_MODELS = set([SqlRegisteredModel.name.key, SqlRegisteredModel.last_updated_time.key])
```

File: mlflow/store/model_registry/sqlalchemy_store.py [hunk 4]
```diff
@@ -297,6 +296,33 @@ def compute_next_token(current_size):
...
+        of order_bys. Registered models are naturally ordered first by name ascending.
+        """
+        clauses = []
+        if order_by_list:
+            for order_by_clause in order_by_list:
+                attribute_token, ascending = SearchUtils.parse_order_by_registered_models(
+                    order_by_clause)
+                if attribute_token == 'name':
+                    field = SqlRegisteredModel.name
+                elif attribute_token == 'last_updated_timestamp':
```
Comment (mparkhe) [inline mlflow/store/model_registry/sqlalchemy_store.py:311]: like comment in `search_utils.py` use existing constants.

```
if attribute_token == SqlRegisteredModel.name.key:
  field = SqlRegisteredModel.name
elif attribute_token == SqlRegisteredModel.last_updated_time.key:
  ...
```

File: mlflow/utils/search_utils.py [hunk 5]
```diff
@@ -317,30 +318,43 @@ def run_matches(run):
         return [run for run in runs if run_matches(run)]
 
     @classmethod
-    def parse_order_by(cls, order_by):
+    def _parse_order_by_string(cls, order_by):
```
Comment (mparkhe) [inline mlflow/utils/search_utils.py:321]: Great refactor!

File: tests/store/model_registry/test_sqlalchemy_store.py [hunk 6]
```diff
@@ -580,3 +582,110 @@ def test_search_registered_model_pagination(self):
...
+    def test_search_registered_model_order_by(self):
+        rms = []
+        # explicitly mock the creation_timestamps because timestamps seem to be unstable in Windows
+        for i in range(50):
+            with mock.patch("mlflow.store.model_registry.sqlalchemy_store.now", return_value=i):
+                rms.append(self._rm_maker(f"RM{i:03}").name)
+
+        # test flow with fixed max_results and order_by (test stable order across pages)
+        returned_rms = []
+        query = "name LIKE 'RM%'"
```
Comment (mparkhe) [inline tests/store/model_registry/test_sqlalchemy_store.py:595]: reminder: handle in code and put in tests for `like`, `Like`, `iLike` ... etc

File: mlflow/utils/search_utils.py (hunk 2 above)
Comment (ankitmathur-db) [reply mlflow/utils/search_utils.py:350]: Yeah definitely agree

File: tests/store/model_registry/test_sqlalchemy_store.py (hunk 6 above)
Comment (ankitmathur-db) [reply tests/store/model_registry/test_sqlalchemy_store.py:595]: Tests for these cases are in #2968 - do you feel it should also be tested in conjunction with the `order_by`?

Also, what do you mean by handle in code?

Comment (mparkhe) [reply tests/store/model_registry/test_sqlalchemy_store.py:595]: aah! i see. can can keep that PR separate

File: mlflow/utils/search_utils.py [hunk 7]
```diff
@@ -331,10 +335,23 @@ def _parse_order_by_string(cls, order_by):
             raise MlflowException(f"Invalid order_by clause '{order_by}'. Could not be parsed.",
                                   error_code=INVALID_PARAMETER_VALUE)
         statement = parsed[0]
-        if len(statement.tokens) != 1 or not isinstance(statement[0], Identifier):
+        if len(statement.tokens) == 1 and \
+                (isinstance(statement[0], Identifier)
+                 or statement.tokens[0].match(ttype=TokenType.Keyword, values=["timestamp"])):
```
Comment (mparkhe) [inline mlflow/utils/search_utils.py:340]: use constant for `"timestamp"`

File: mlflow/utils/search_utils.py [hunk 8]
```diff
@@ -331,10 +335,23 @@ def _parse_order_by_string(cls, order_by):
...
+            token_value = statement.tokens[0].value
```
Comment (mparkhe) [inline mlflow/utils/search_utils.py:341]: just use the timestamp constant also in the 3-parts case

File: tests/store/model_registry/test_sqlalchemy_store.py [hunk 9]
```diff
@@ -612,12 +612,30 @@ def test_search_registered_model_order_by(self):
                                                    order_by=['last_updated_timestamp DESC'],
                                                    max_results=100)
         self.assertEqual(rms[::-1], result)
+        # timestamp returns same result as last_updated_timestamp
+        result, _ = self._search_registered_models(query,
+                                                   page_token=None,
+                                                   order_by=['timestamp DESC'],
```
Comment (mparkhe) [inline tests/store/model_registry/test_sqlalchemy_store.py:618]: for for different flavors of blanks
- space (already covered)
- multiple spaces
- tabs

File: mlflow/utils/search_utils.py [hunk 10]
```diff
@@ -26,8 +26,12 @@ class SearchUtils(object):
         CASE_INSENSITIVE_STRING_COMPARISON_OPERATORS.union({'='})
     VALID_SEARCH_ATTRIBUTE_KEYS = set(RunInfo.get_searchable_attributes())
     VALID_ORDER_BY_ATTRIBUTE_KEYS = set(RunInfo.get_orderable_attributes())
-    VALID_ORDER_BY_KEYS_REGISTERED_MODELS = set([SqlRegisteredModel.name.key,
-                                                 SqlRegisteredModel.last_updated_time.key])
+    VALID_TIMESTAMP_ORDER_BY_KEYS = set(["timestamp", "last_updated_timestamp"])
+    VALID_ORDER_BY_KEYS_REGISTERED_MODELS = set([SqlRegisteredModel.name.key])\
+        .union(VALID_TIMESTAMP_ORDER_BY_KEYS)
+    # We encourage users to use timestamp for order-by
+    RECOMMENDED_ORDER_BY_KEYS_REGISTERED_MODELS = VALID_ORDER_BY_KEYS_REGISTERED_MODELS\
```
Comment (mparkhe) [inline mlflow/utils/search_utils.py:33]: - create a constant for `"timestamp"`
- replace `RECOMMENDED_ORDER_BY_KEYS_REGISTERED_MODELS` with something like this

```
DOCUMENTED_ORDER_BY_KEYS_FOR_REGISTERED_MODELS = [ORDER_BY_KEY_MODEL_NAME, ORDER_BY_KEY_TIMESTAMP]
```

**Review by mparkhe — COMMENTED** [review]
some small nits

**Review by mparkhe — APPROVED** [review]
Looks good. In a follow up PR please add a unit test for `Asc`, `asc`, `desc` ... etc.

**Pull request merged**